### ISSUE TITLE
Custom image as background for SSO?

### ISSUE THREAD

**Issue by eauchat** [YunoHost/issues#1267]
On my yunohost install, I changed manually some CSS styles and added a custom picture as background to the SSO, it makes it look really nice. I thought it could be nice that this could be achieved easily, maybe through the admin interface.  
From what I imagine, this would include:
  - make the UI in yuohost-admin,
  - make the system to upload images,
  - choose where to upload it to (is it to yunohost-admin? or does this location risk deletion on updates?)

Another way could be to implement it per user, similar to user's avatar picture #879 it could be put in the LDAP user's profile.

**Comment by Psycojoker** [comment]
Hello,

This is the kind of idea we discuss from time to time but never got the time to work on, we just never open a ticket for that but that would definitively be nice to have that one day ™

A problem is that the SSOwat code base is not a fun part of the code on which to work so we aren't really motivated to put time into it :/

Thx for taking the time to suggest this idea :)

**Comment by eauchat** [comment]
I forgot to mention posting the issue that I would be interested in working on this. Together with #879. But I didn't look close to the SSO code yet.

I think the two things that I'm not so clear about are:
 - what would be a good way/tool to uploading images from SSO or/and admin panel UI (understand what I mean? ;/)
 - is it possible to simply upload the image to /usr/share/ssowat/assets/img or could it risk deletion on an update? If yes, what better alternative?

**Comment by eauchat** [comment]
@isAAAc
So the way I did it was to add the background image I wanted in `/usr/share/ssowat/portal/assets/img/` , and then to change styles in `/usr/share/ssowat/portal/assets/css/ynh-style.css`  to add a background-image to body.
From `ynh-style.css`  and ynhpanel.css you can customize ssowat display, for example also change the yunohost logo, or other things.
It's quite cool, not at all complicated to do if you know basic CSS.
I just don't know if it will last when I will update yunohost.

**Comment by alexAubin** [comment]
This would definitely be a pretty cool feature :+1: Imho this is heavily related to SSO theming in general (so for instance changing the 'Yunohost' logo to your club/association logo on the SSO portal)

>I think the two things that I'm not so clear about are:
>    - what would be a good way/tool to uploading images from SSO or/and admin panel UI (understand what I mean? ;/)
>    - is it possible to simply upload the image to /usr/share/ssowat/assets/img or could it risk deletion on an update? If yes, what better alternative?

Naively I think the whole upload thing is probably going to be pretty complicated. I think a first goal would be to have a system where this can at least be handled via CLI / manual tweaks which are clean and persistent (compared to, say, manually editing raw css files).

Random ideas but maybe having an `override` folder which, if it contains some file like `background.png`, would then be used instead of the default thing ? (I have basically no experience with designing theme-mechanism so I dunno what's a good design pattern-thing for this)

**Comment by eauchat** [comment]
I think this idea of making theming easy is great.

I see two levels in that:
  - one would be the ability to easily install themes made by people, the same way you'd install packages (like a packet manager for themes)
  - customize easily any current theme you're using with your own values: like background image, yunohost logo image, or even add some CSS.

The way I see it:
  - create the directory: `ssowat/portal/assets/themes` This directory should be respected on updates (the system should never modify it's content).
  - in that directory, every folder is a theme
  - in the admin panel there is one more "menu": theming, and from this menu you can:
    * browse list of installed themes (corresponding to directories in `ssowat/portal/assets/themes`) and select the active one,
    * search new themes with a theme package manager
    * edit any theme to add your custom needs.
All this could look similar to the apps admin panel and so would the corresponding CLI.

The thing I'm not so sure in this architecture is how the custom theming is behaving.
Is it one more styling sheet on top of default styles and themed styles, or is it just the ability to easily edit a theme.
In other words, is it important to easily change theme whilst keeping some parameters (like background image and logo image values). Or can you simply customize any theme easily, and if you switch theme and want to keep your custom association logo, you just reupload it in the new theme you chose.
I feel the second is simpler to implement and gives more power to the administrator.

Hope all this is understandable :)

**Comment by alexAubin** [comment]
Closing because themes have been merged
